src/box/xrow_update_array.c

```diff
@@ -357,7 +357,11 @@ xrow_update_op_do_array_insert(struct xrow_update_op *op,
 		return xrow_update_op_do_field_insert(op, &item->field);
 	}
 
-	if (xrow_update_array_append_nils(field, op) != 0)
+	/*
+	 * Allow autofill with nulls only for root array
+	 */
+	if (op->lexer.src == NULL &&
```

Comment:
This better be moved into append_nils in case we see append is necessary. Otherwise you added a branch for the operations which don't need the append anyway (which is the most common case).